File: capy/src/e2e-cfun/java/dev/capylang/test/EnumCollectionTest.java
package dev.capylang.test;

import capy.lang.Result;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class EnumCollectionTest {
    @Test
    void matchCanCoverAllEnumValues() {
        assertThat(EnumCollection.colorMatch(EnumCollection.Color.RED)).isEqualTo(1);
        assertThat(EnumCollection.colorMatch(EnumCollection.Color.BLUE)).isEqualTo(2);
        assertThat(EnumCollection.colorMatch(EnumCollection.Color.GREEN_BLUE)).isEqualTo(3);
        assertThat(EnumCollection.colorMatch(EnumCollection.Color.YELLOW)).isEqualTo(4);
        assertThat(EnumCollection.colorMatch(EnumCollection.Color.BLACK)).isEqualTo(5);
        assertThat(EnumCollection.colorMatch(EnumCollection.Color.WHITE)).isEqualTo(6);
    }

    @Test
    void matchCanUseWildcardForEnums() {
        assertThat(EnumCollection.colorMatchWildcard(EnumCollection.Color.RED)).isEqualTo(1);
        assertThat(EnumCollection.colorMatchWildcard(EnumCollection.Color.BLUE)).isEqualTo(2);
        assertThat(EnumCollection.colorMatchWildcard(EnumCollection.Color.WHITE)).isEqualTo(0);
    }

    @Test
    void enumProvidesOrderAndName() {
        assertThat(EnumCollection.colorOrder(EnumCollection.Color.RED)).isEqualTo(0);
        assertThat(EnumCollection.colorOrder(EnumCollection.Color.GREEN_BLUE)).isEqualTo(2);
        assertThat(EnumCollection.colorName(EnumCollection.Color.GREEN_BLUE)).isEqualTo("GREEN_BLUE");
    }

    @Test
    void valuesReturnsEnumSet() {
        assertThat(EnumCollection.colorValues())
                .containsExactlyInAnyOrder(EnumCollection.Color.values());
    }

    @Test
    void parseFromStringAndOrderReturnsResult() {
        assertThat(EnumCollection.parseColorName("BLUE"))
                .isInstanceOf(Result.Success.class);
        assertThat(((Result.Success<EnumCollection.Color>) EnumCollection.parseColorName("BLUE")).value())
                .isEqualTo(EnumCollection.Color.BLUE);

        assertThat(EnumCollection.parseColorOrder(3))
                .isInstanceOf(Result.Success.class);
        assertThat(((Result.Success<EnumCollection.Color>) EnumCollection.parseColorOrder(3)).value())
                .isEqualTo(EnumCollection.Color.YELLOW);

        assertThat(EnumCollection.parseColorName("UNKNOWN"))
                .isInstanceOf(Result.Error.class);
        assertThat(EnumCollection.parseColorOrder(99))
                .isInstanceOf(Result.Error.class);
    }

    @Test
    void enumValuesCanBeUsedWithoutBracesInExpressions() {
        assertThat(EnumCollection.qualifiedColorWithoutBraces())
                .isEqualTo(EnumCollection.Color.BLUE);
        assertThat(EnumCollection.unqualifiedColorWithoutBraces())
                .isEqualTo(EnumCollection.Color.GREEN_BLUE);
        assertThat(EnumCollection.inferredColorWithoutBraces())
                .isEqualTo(EnumCollection.Color.RED);
        assertThat(EnumCollection.colorNameWithoutBraces())
                .isEqualTo("YELLOW");
    }

    @Test
    void enumValuesCanBeUsedWithoutBracesInCompoundExpressions() {
        assertThat(EnumCollection.colorListWithoutBraces())
                .containsExactly(
                        EnumCollection.Color.RED,
                        EnumCollection.Color.BLUE,
                        EnumCollection.Color.GREEN_BLUE);

        var palette = EnumCollection.paletteWithoutBraces();
        assertThat(palette.primary()).isEqualTo(EnumCollection.Color.WHITE);
        assertThat(palette.fallback()).isEqualTo(EnumCollection.Color.BLACK);
    }
}

File: capy/src/e2e-cfun/js/dev/capylang/test/EnumCollection.test.js
const test = require('node:test');
const { assert, assertCapyEqual, assertError, assertSuccess, generatedModule } = require('./_support');

test('EnumCollection', () => {
    const enumCollection = generatedModule('dev/capylang/test/EnumCollection.js');
    assert.equal(enumCollection.colorMatch(enumCollection.RED), 1);
    assert.equal(enumCollection.colorMatch(enumCollection.GREEN_BLUE), 3);
    assert.equal(enumCollection.colorMatchWildcard(enumCollection.WHITE), 0);
    assert.equal(enumCollection.colorOrder(enumCollection.GREEN_BLUE), 2);
    assert.equal(enumCollection.colorName(enumCollection.GREEN_BLUE), 'GREEN_BLUE');
    assertCapyEqual(enumCollection.colorValues(), new Set([
        enumCollection.RED,
        enumCollection.BLUE,
        enumCollection.GREEN_BLUE,
        enumCollection.YELLOW,
        enumCollection.BLACK,
        enumCollection.WHITE,
    ]));
    assert.equal(assertSuccess(enumCollection.parseColorName('BLUE')), enumCollection.BLUE);
    assert.equal(assertSuccess(enumCollection.parseColorOrder(3)), enumCollection.YELLOW);
    assert.match(assertError(enumCollection.parseColorName('UNKNOWN')), /Unable to parse Color/);
    assert.equal(enumCollection.qualifiedColorWithoutBraces(), enumCollection.BLUE);
    assert.equal(enumCollection.unqualifiedColorWithoutBraces(), enumCollection.GREEN_BLUE);
    assert.equal(enumCollection.inferredColorWithoutBraces(), enumCollection.RED);
    assertCapyEqual(enumCollection.colorListWithoutBraces(), [enumCollection.RED, enumCollection.BLUE, enumCollection.GREEN_BLUE]);
    assert.equal(enumCollection.paletteWithoutBraces().primary, enumCollection.WHITE);
});
